src/lib/common/sol-platform-linux-common.c

```diff
@@ -263,3 +307,581 @@ sol_platform_impl_get_os_version(char **version)
+{
+    int fd = 0;
+    struct sol_vector lines;
+    struct sol_str_slice content = SOL_STR_SLICE_EMPTY;
+    struct sol_str_slice *itr;
+    struct sol_file_reader *reader;
+    uint16_t idx;
+
+    fd = open(SOL_MTAB_FILE, O_RDONLY | O_CLOEXEC);
+    if (!fd) {
+        SOL_ERR("Could not open "SOL_MTAB_FILE " - %s",
+            sol_util_strerrora(errno));
+        return -errno;
+    }
+
+    reader = sol_file_reader_fd(fd);
+    if (!reader) {
+        SOL_ERR("Could not read "SOL_MTAB_FILE " file - %s",
+            sol_util_strerrora(errno));
+        return -errno;
+    }
+
+    content = sol_file_reader_get_all(reader);
+    lines = sol_util_str_split(content, "\n", 0);
+    SOL_VECTOR_FOREACH_IDX (&lines, itr, idx) {
+        char buff[512];
+        snprintf(buff, itr->len, "%s", itr->data);
+        if (!strstr(buff, mpoint))
+            sol_buffer_append_printf(output, "%s", buff);
+    }
+
+    sol_vector_clear(&lines);
+    sol_file_reader_close(reader);
+    close(fd);
+    return 0;
+}
+
+// since we don't have a mtab entry remove function we must implement it ourselves
+static int
+sol_mtab_cleanup(const char *mpoint)
+{
+    int res, err, fd;
+    struct sol_buffer output;
+
+    res = fd = 0;
+    sol_buffer_init(&output);
+
+    err = sol_mtab_remove_entry(mpoint, &output);
+    SOL_INT_CHECK(err, < 0, err);
+
+    fd = open(SOL_MTAB_FILE, O_RDWR | O_CLOEXEC | O_TRUNC);
+    if (!fd) {
+        SOL_ERR("Could not open " SOL_MTAB_FILE);
+        return -errno;
+    }
+
+    err = write(fd, output.data, sizeof(output.data));
+    if (err < -1) {
+        SOL_ERR("Could not write "SOL_MTAB_FILE " file - %s",
+            sol_util_strerrora(errno));
+        res = -errno;
+    }
+
+    err = close(fd);
+    if (err < -1) {
+        SOL_ERR("Could not close "SOL_MTAB_FILE " file - %s",
+            sol_util_strerrora(errno));
+        res = -errno;
+    }
+
+    sol_buffer_fini(&output);
+    return res;
+}
+
+static int
+run_umount(const char *mpoint)
+{
+    int err;
+
+    err = sol_mtab_cleanup(mpoint);
+    SOL_INT_CHECK(err, < 0, err);
+
+    err = umount(mpoint);
+    if (err != 0) {
+        SOL_ERR("Couldn't umount %s - %s", mpoint, sol_util_strerrora(errno));
+        return -errno;
+    }
+
+    return err;
+}
+
+static void
+on_umount_fork(void *data)
+{
+    int err;
+    struct umount_data *udata = data;
+
+    err = run_umount(udata->mpoint);
+    sol_platform_linux_fork_run_exit(err);
+}
+
+static void
+on_umount_fork_exit(void *data, uint64_t pid, int status)
+{
+    struct umount_data *udata = data;
+
+    udata->async_cb(udata->mpoint, udata->data, pid, status);
+    free(udata->mpoint);
+    free(udata);
+}
+
+int
+sol_platform_impl_umount(const char *mpoint, void (*async_cb)(const char *mpoint, void *data, uint64_t pid, int status), void *data)
+{
+    struct umount_data *udata;
+
+    SOL_NULL_CHECK(mpoint, -EINVAL);
+
+    if (!async_cb)
+        return run_umount(mpoint);
```

Comment:
avoid doing this kind of widely different behavior. If you find another case like this in the future, keep the fork-run (or thread, or idler/timer) flow and just skip the callback at the end, it will keep the execution flow the same, avoiding problems.